include multiple keys from a single backup

martgil opened the pull request:
This PR will:

- brings back the green backup button for accounts with multiple private keys but the other doesn't have backup.
- backup multiple keys in a single file when the multiple private keys are detected but few are only having backup.

close #2482 // if this PR closes an issue

issue #2806 // if it doesn't close the issue yet

Task list:

- [x] backups should contain multiple keys in single file
- [ ] all protected by the same passphrase. (uncertain. At the moment, I use the passphrase of key[0] which I believe the active key of the user to backup all the keys)
- [x] Let user uncheck keys they don't want to back up
- [x] Fix button not currently shown.
- [ ] Missing tests for checkboxes; submitted keys after performing a successful backup
----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests will be added later (issue #...)
- Tests added or updated

** Incomplete functionalities from issue #2482
- [ ] Get currently used key and it's associated passphrase when encrypting multiple backup keys.
- [x] Render a list of backup (probably a checkboxes) for users to include/exclude keys

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

martgil commented:
I am really uncertain of this line code. This line of code should be responsible for getting the current primary key being used by the logged-in user. But here I have temporarily use the first primary key which I'm totally uncertain if was really correct by encrypting the backup using the passphrase of the first primary key found.

https://github.com/FlowCrypt/flowcrypt-browser/blob/02137ede6ac4d54ff5f38196eb93966b27926b6d/extension/chrome/settings/modules/backup-manual-module.ts#L82

martgil commented:
Can I also ask for a suggestion on how to generally approach fixing CI-related issues especially on "GMAIL Live tests"?

I already tried running Gmail live tests on my local machine but I think I have missed something. 
```
npm run test_ci_chrome_consumer_live_gmail

flowcrypt-browser@8.1.4 test_ci_chrome_consumer_live_gmail
node ./node_modules/ava/cli --timeout=20m --verbose --concurrency=1 build/test/test/source/test.js -- CONSUMER-LIVE-GMAIL STANDARD-GROUP

TEST_VARIANT: CONSUMER-LIVE-GMAIL:STANDARD-GROUP, (build dir: build/chrome-consumer, poolSizeOne: false)
consts:  {"TIMEOUT_SHORT":60000,"TIMEOUT_EACH_RETRY":180000,"TIMEOUT_ALL_RETRIES":780000,"TIMEOUT_OVERALL":840000,"ATTEMPTS":3,"POOL_SIZE":3,"IS_LOCAL_DEBUG":false} 

  ✖ No tests found in build/test/test/source/test.js
  ```

limonte commented:
> Can I also ask for a suggestion on how to generally approach fixing CI-related issues especially on "GMAIL Live tests"?
> 
> I already tried running Gmail live tests on my local machine but I think I have missed something.
> 
> ```
> npm run test_ci_chrome_consumer_live_gmail
> 
> flowcrypt-browser@8.1.4 test_ci_chrome_consumer_live_gmail
> node ./node_modules/ava/cli --timeout=20m --verbose --concurrency=1 build/test/test/source/test.js -- CONSUMER-LIVE-GMAIL STANDARD-GROUP
> 
> TEST_VARIANT: CONSUMER-LIVE-GMAIL:STANDARD-GROUP, (build dir: build/chrome-consumer, poolSizeOne: false)
> consts:  {"TIMEOUT_SHORT":60000,"TIMEOUT_EACH_RETRY":180000,"TIMEOUT_ALL_RETRIES":780000,"TIMEOUT_OVERALL":840000,"ATTEMPTS":3,"POOL_SIZE":3,"IS_LOCAL_DEBUG":false} 
> 
>   ✖ No tests found in build/test/test/source/test.js
> ```

The problem with Gmail tests is unrelated to this PR and has been fixed. I re-run tests for this PR, they are passing now.

martgil commented:
Thanks a lot for taking a look, @limonte. The tests are passing now.

Kind Regards,
Mart Gil

limonte commented on extension/chrome/settings/modules/backup-status-module.ts:
```diff
@@ -81,10 +80,42 @@ export class BackupStatusModule extends ViewModule<BackupView> {
+      $('.key_backup_selection').append(dom);
+    }
+    $('.input_prvkey_backup_checkbox').click((event) => {
+      const email = String($($($(event.target).siblings()[0]).children()[0]).children()[0].innerText);
```
Instead of this [spaghetti code](https://en.wikipedia.org/wiki/Spaghetti_code) I'd recommend this:

```html
<input class="input_prvkey_backup_checkbox" type="checkbox" data-email="${primaryKi.emails}"/>
```

```js
const email = $(event.target).data('email');
```

rrrooommmaaa commented:
Hi @martgil This is what I see
![image](https://user-images.githubusercontent.com/23335944/134805696-fd37f1ec-1e1c-430d-9b8c-b8b6b85c7e3c.png)

We usually display spaced fingerprints (using `Str.spaced()`) -- you can have a look at contacts
![image](https://user-images.githubusercontent.com/23335944/134805903-dd0933cf-6c53-49fe-9f68-58248f8100cc.png)
We should render only primary key's fingerprint (similar to what `that's id field) should be enough
As x509 backup isn't currently supported by https://github.com/FlowCrypt/flowcrypt-browser/issues/2482#issuecomment-922475768, I suggest to make them un-clickable (disabled) and add a field of type along with the fingerprint.
Something like: [] - openpgp - email - fingerprint,

rrrooommmaaa commented:
ake

rrrooommmaaa commented:
@martgil I have added a stub unit test for backup page -- `settings - manual backup several keys`.
Please pull and fill with checks.

rrrooommmaaa commented:
and after saving to file, we can use `awaitDownloadTriggeredByClicking` to check the contents of the generated file.
As I understand all OpenPGP keys should be under the same armor, so `parseMany` will return sevral keys.

martgil commented:
Thank you sir @rrrooommmaaa for adding a test to start with. I added the checks for verifying the submitted back up and they are passing with the tests I've added. I use the `KeyUtil.readMany` instead of `OpenPGP.parseMany` at the moment because `OpenPGP.parseMany` gives me only 1 armored key while `KeyUtil.readMany` returns all armored key. Several tests are still failing maybe because of the merge origin/master because I tried to fix the tests errors on `issue-3885-smart-alias-submission` that I'm getting similar to this PR. I'm still looking into it and checking the code if there's anything I could improve. Thanks again, sir!

rrrooommmaaa commented:
Hi @martgil The failing test is not realted to your updates, the guys are working on how to fix it.

>  because OpenPGP.parseMany gives me only 1 armored key

Is this because each key has its own armor header/footer in the file?
like
```
-----BEGIN PGP PRIVATE KEY BLOCK-----
....
-----END PGP PRIVATE KEY BLOCK-----
-----BEGIN PGP PRIVATE KEY BLOCK-----
....
-----END PGP PRIVATE KEY BLOCK-----
```

I guess @tomholub meant to armor all the keys under a single armor (as we did in PR #3915 test armoring)  -- thus the single-passphrase requirement makes sense, also easier to extract in some cases. Let's double check with him.

tomholub commented:
> -----BEGIN PGP PRIVATE KEY BLOCK-----
> ....
> -----END PGP PRIVATE KEY BLOCK-----
> -----BEGIN PGP PRIVATE KEY BLOCK-----
> ....
> -----END PGP PRIVATE KEY BLOCK-----

This is the format I was looking for - just concatenated armored blocks. `KeyUtil.readMany` is the right one to use in that case.

I figured this would make it a bit easier if an advanced user wanted to extract a particular key from the backup.. or just visually know what's going on (visually obvious that it's more than one key in the file). But mostly, it's just for consistency with what our apps on other platforms do.

martgil commented:
Hi @rrrooommmaaa and @tomholub.

>Is this because each key has its own armor header/footer in the file?

I think yes. Because when I use the `.join('\n')` in most cases during the key backup preparation. And then, I use the `KeyUtil.readMany` to read fingerprint on checking. And the passphrase I use to "encrypt" the keys is the first `acctEmail` or the initial account I should say (needs verification/clarification).

rrrooommmaaa commented on extension/chrome/settings/modules/backup.ts:
```diff
@@ -15,7 +15,10 @@ import { BackupManualActionModule as BackupManualModule } from './backup-manual-
 import { Lang } from '../../../js/common/lang.js';
 import { AcctStore, EmailProvider } from '../../../js/common/platform/store/acct-store.js';
-
+export interface PrvKeyIdentity {
```
It's a good point to have a `KeyIdentity` type (likely in `extension/js/common/core/crypto/key.ts`).
There are already a couple of places in the codebase that utilize a structure like `{ id: string, type: 'openpgp' | 'x509' }` -- that's the perfect candidate to be `KeyIdentity`. There are plenty of places in the codebase that use `fingerprints[0]`, it's hard to follow whether they assume `fingerprints[0]` to be unique within the retrieved records or not. I see some potential problems, should we happen have (although a very rare case) both `x509` and `openpgp` keys with the same fingerprint -- for example, with different passphrases. It might be worthwhile to introduce functions that return explicitly typed identity, and gradually phase out usage of `fingerprints[0]` where appropriate.
Are you ok to go ahead with such refactoring, @tomholub

martgil replied:
>for example, with different passphrases. It might be worthwhile to introduce functions that return explicitly typed identity, and gradually phase out usage of fingerprints[0] where appropriate.

Thank you, sir! I just move/rename and use the KeyIdentity type. I wasn't able to follow the preceding comments yet in this thread. But I managed to add the recent request change.

rrrooommmaaa commented on test/source/tests/settings.ts:
```diff
@@ -519,6 +520,46 @@ export let defineSettingsTests = (testVariant: TestVariant, testWithBrowser: Tes
+      const backupPage = await browser.newPage(t, TestUrls.extension(`/chrome/settings/modules/backup.htm?acctEmail=${acctEmail}&action=backup_manual&parentTabId=0`));
+      // play with backup controls here
+      expect(await backupPage.isDisabled('[data-fingerprints="47FB03183E03A8ED44E3BBFCCEA2D53BB9D24871"]')).to.equal(true);
+      expect(await backupPage.isDisabled('[data-fingerprints="5A08466253C956E9C76C2E95A35068FD4E037879"]')).to.equal(true);
```
They are disabled ok, but why are they checked? They should be unchecked as they won't be included in the backup
![image](https://user-images.githubusercontent.com/23335944/135438017-8aa36e6a-360d-42dd-b450-85d8067dfd13.png)

martgil replied:
Thank you, sir. They are now unchecked and disabled by now.

rrrooommmaaa commented on test/source/tests/settings.ts:
```diff
@@ -519,6 +520,46 @@ export let defineSettingsTests = (testVariant: TestVariant, testWithBrowser: Tes
+      // play with backup controls here
+      expect(await backupPage.isDisabled('[data-fingerprints="47FB03183E03A8ED44E3BBFCCEA2D53BB9D24871"]')).to.equal(true);
+      expect(await backupPage.isDisabled('[data-fingerprints="5A08466253C956E9C76C2E95A35068FD4E037879"]')).to.equal(true);
+      await backupPage.waitAndClick('[data-fingerprints="CB0485FE44FC22FF09AF0DB31B383D0334E38B28,B38FC17737BB1D66C69CB25975796017BA7823CF"]'); // uncheck
```
Please also test that these two are initially checked

martgil replied:
I've added a test to detect if these two checkboxes are "unchecked".

rrrooommmaaa commented on test/source/tests/settings.ts:
```diff
@@ -519,6 +520,46 @@ export let defineSettingsTests = (testVariant: TestVariant, testWithBrowser: Tes
+      expect(!keys[0].allIds.some(fingerprint => ['CB0485FE44FC22FF09AF0DB31B383D0334E38B28', 'B38FC17737BB1D66C69CB25975796017BA7823CF'].includes(fingerprint))).to.equal(true);
+      await backupPage.waitAndRespondToModal('info', 'confirm', 'Downloading private key backup file');
+      await backupPage.waitAndRespondToModal('info', 'confirm', 'Your private key has been successfully backed up');
+      await backupPage.close();
```
Please also include a test of backing up more than one key (protected in `KeyStore` with different passphrases).

rrrooommmaaa commented on extension/chrome/settings/modules/backup-manual-module.ts:
```diff
@@ -53,17 +54,78 @@ export class BackupManualActionModule extends ViewModule<BackupView> {
+      `.trim();
+      $('.key_backup_selection').append(dom); // xss-escaped
+      if (keyType !== 'x509') {
+        this.addKeyToBackup({ 'email': email, 'fingerprints': fingerprints });
```
no need to quote property names. In this case `{ email, fingerprints }` should do, I think.

martgil replied:
It works! Thank you!

rrrooommmaaa commented on extension/chrome/settings/modules/backup-manual-module.ts:
```diff
@@ -53,17 +54,78 @@ export class BackupManualActionModule extends ViewModule<BackupView> {
+
+  private renderPrvKeysBackupSelection = async (primaryKeys: ExtendedKeyInfo[]) => {
+    for (const primaryKi of primaryKeys) {
+      const email = Xss.escape(String(primaryKi.emails));
```
This is how it looks now.
![image](https://user-images.githubusercontent.com/23335944/135439316-6ad7c7f6-41ec-49cb-a557-1ebdce83a0a1.png)
@martgil Do you really want to display all emails the key may have? I would use `primaryKi.emails.join()` or `primaryKi.emails.toString()` for extra build-time type check -- this is why we're using Typescript after all. Have you tested how it would look visually with more than one email?
I guess `email[0]` is enough, @tomholub ?

martgil replied:
> Have you tested how it would look visually with more than one email?

Yes, sir! I make the container available for scrolling when it reaches the maximum height of 300px.

rrrooommmaaa commented on extension/chrome/settings/modules/backup.ts:
```diff
@@ -95,9 +101,9 @@ export class BackupView extends View {
   }
 
-  public setHandlers = () => {
+  public setHandlers = async () => {
```
making this function `async` means that `View.setTestViewStateLoaded();` may be called prior to its completion.
I modified `extension/js/common/view.ts` to await its completion in https://github.com/FlowCrypt/flowcrypt-browser/pull/3973/commits/9b1aa9fa55b51c43b2f1d79617b9bb3e1b6cab54, so should be good now.
Wasn't it appropriate to have this logic in `render()`? -- that's the only `async` method.

martgil replied:
>Wasn't it appropriate to have this logic in render()? -- that's the only async method.

You're right sir. It's really ugly to put some async function there. I just move those related functions to `backup.ts` itself. Thank you for your patience, sir!

martgil commented on extension/chrome/settings/modules/backup-manual-module.ts:
```diff
@@ -53,17 +54,78 @@ export class BackupManualActionModule extends ViewModule<BackupView> {
+            <input class="input_prvkey_backup_checkbox" type="checkbox" checked data-emails="${email}" data-fingerprints="${fingerprints}" ${keyType === 'x509' ? 'disabled' : ''} />
+            ${email}
+            </p>
+            <p class="m-0 prv_fingerprint"><span>${keyType} - ${email} - ${Str.spaced(fingerprints[0])}</span></p>
```
Okay sir, ill change that. Thanks!

martgil commented:
Hello sir, @rrrooommmaaa. I think it is now ready for review.

rrrooommmaaa commented on test/source/tests/settings.ts:
```diff
@@ -519,6 +520,54 @@ export const defineSettingsTests = (testVariant: TestVariant, testWithBrowser: T
+        { isSavePassphraseChecked: true, isSavePassphraseHidden: false });
+      const backupPage = await browser.newPage(t, TestUrls.extension(`/chrome/settings/modules/backup.htm?acctEmail=${acctEmail}&action=backup_manual&parentTabId=0`));
+      // OpenPGP keys are checked, x509 keys are unchecked
+      expect(await backupPage.isChecked('[data-id="47FB03183E03A8ED44E3BBFCCEA2D53BB9D24871"]')).to.equal(false);
```
@limonte Why didn't we ever use `data-` in this project before? I see we have always been using attrs. What would you recommend?

rrrooommmaaa commented on extension/css/cryptup.css:
```diff
@@ -44,6 +44,7 @@ button.swal2-styled:focus,
   text-decoration: none;
```
@limonte are CSS updates in this file looking good to you?

limonte replied:
yup, all good 👍
